### PR TITLE
Update sensor fusion

### PR DESCRIPTION
**Pull request by antimonyB**

### CODE REVIEW

File: Autopilot/AttitudeManager/Inc/AttitudeDatatypes.hpp
Comment (Pop0097) [inline Autopilot/AttitudeManager/Inc/AttitudeDatatypes.hpp]: I think it would be best to just do `struct STRUCT_NAME` instead of using `typedef` to keep consistency with the other class declarations.

File: Autopilot/AttitudeManager/Inc/AttitudeDatatypes.hpp
Comment (Pop0097) [inline Autopilot/AttitudeManager/Inc/AttitudeDatatypes.hpp]: The name for this needs to be changed because it is the same as the one declared in altimeter.hpp

maybe use `Altimeter_Data_t`

File: Autopilot/AttitudeManager/Inc/AttitudeDatatypes.hpp
Comment (Pop0097) [inline Autopilot/AttitudeManager/Inc/AttitudeDatatypes.hpp]: Same as other comment. Use `GPS_Data_t`

File: Autopilot/Inc/SensorFusion.hpp
Comment (Pop0097) [inline Autopilot/Inc/SensorFusion.hpp]: I don't think `typedef` is needed here. With C++ structs there really is no need

File: Autopilot/Inc/SensorFusion.hpp
Comment (Pop0097) [inline Autopilot/Inc/SensorFusion.hpp]: When you change the struct names in AttitudeDatatypes.hpp, remember to change them over here

File: Autopilot/Inc/SensorFusion.hpp
Comment (Pop0097) [inline Autopilot/Inc/SensorFusion.hpp]: Perhaps use a more descriptive name for this function? The other function is called SF_GetPosition.

Perhaps, since SF_GetResult is used by Attitude Manager and SF_GetPosition is used by Path Manager, we could name them SF_AttitudeManager_Result and SF_PathManager_Result? It'll also make it clear to future people which parts of our code call each function.

**Review by Pop0097 — CHANGES_REQUESTED** [review]
Looks great so far! Just gotta write some unit tests to make sure the code works for AttitudeManager and PathManager!

**Comment by Pop0097** [comment]
Some things I forgot to mention: Take a look at the travis tests. There are some edits to CMakeLists.txt that you will need to make

Errors seem to be that the compiler cannot find SensorFusion.hpp and "arm_math.h".

File: Autopilot/AttitudeManager/Inc/AttitudeDatatypes.hpp
Comment (Fberbari) [inline Autopilot/AttitudeManager/Inc/AttitudeDatatypes.hpp]: no need for this.

File: Autopilot/Inc/SensorFusion.hpp
Comment (Fberbari) [inline Autopilot/Inc/SensorFusion.hpp]: SFAttitudeOutput vs SFPathOutput would be better names

File: Autopilot/Inc/SensorFusion.hpp
Comment (Fberbari) [inline Autopilot/Inc/SensorFusion.hpp]: Probs best stored in the cpp file rather than the interface file.

File: Autopilot/Src/SensorFusion.cpp
Comment (Fberbari) [inline Autopilot/Src/SensorFusion.cpp]: zwe should use a define from up in CMake that contains the running frequency of the processor.

File: Autopilot/Inc/SensorFusion.hpp
Comment (Pop0097) [inline Autopilot/Inc/SensorFusion.hpp]: Will you be outputting a heading?

File: Autopilot/Inc/gps.hpp
Comment (Pop0097) [inline Autopilot/Inc/gps.hpp]: `static TestGps* GetInstance() {}`

File: Autopilot/Test/Src/TestGps.cpp
Comment (Pop0097) [inline Autopilot/Test/Src/TestGps.cpp]: I would make a new directry in the `./Test/Src` folder called `./Test/Src/SensorFusion`. You can put all sensor test files and the sensor fusion test file within.

Also, for consistent file naming, please use `Test_SENSORNAME.cpp`

File: Autopilot/Test/Src/Test_SF.cpp
Comment (Pop0097) [inline Autopilot/Test/Src/Test_SF.cpp]: Will need to develop a set of tests so we can be sure future edits don't break your code. Probably will keep this for your next merge since we are short on time though

File: Autopilot/CMakeLists.txt
Comment (Pop0097) [inline Autopilot/CMakeLists.txt]: `sensorFusionModules`. Also use `SENSOR_FUSION_MODULES_...`

File: Autopilot/CMakeLists.txt
Comment (Pop0097) [inline Autopilot/CMakeLists.txt]: Will anything need to be added to the Simulation build part of the cmake?

File: Autopilot/SensorFusion/Src/SensorFusion.cpp
Comment (Pop0097) [inline Autopilot/SensorFusion/Src/SensorFusion.cpp]: should be `imu.hpp`

File: Autopilot/SensorFusion/Src/SensorFusion.cpp
Comment (Pop0097) [inline Autopilot/SensorFusion/Src/SensorFusion.cpp:43]: space between multiplication sign

File: Autopilot/SensorFusion/Src/SensorFusion.cpp
Comment (Pop0097) [inline Autopilot/SensorFusion/Src/SensorFusion.cpp:28]: is `typedef` needed here?

File: Autopilot/CControl/Inc/CControlFunctions.h
Comment (Pop0097) [inline Autopilot/CControl/Inc/CControlFunctions.h:6]: Are these verified to work?

File: Autopilot/SensorFusion/Src/SensorFusion.cpp
Comment (Pop0097) [inline Autopilot/SensorFusion/Src/SensorFusion.cpp:71]: I'd refrain from using one-character variable names. It's harder to understand

File: Autopilot/SensorFusion/Src/SensorFusion.cpp
Comment (Pop0097) [inline Autopilot/SensorFusion/Src/SensorFusion.cpp:105]: This is a bit dangerous because if we return anything other than 0 the plane will switch over to safety... If you could come up with a list of error codes for sensor fusion and let us know which ones should cause fatal behaviour, that would be great.

File: Autopilot/SensorFusion/Src/SensorFusion.cpp
Comment (Pop0097) [inline Autopilot/SensorFusion/Src/SensorFusion.cpp:112]: Same thing here. We need to decide on a set of non-fatal error codes. 

Also, I think enums for these would be very appreciated. But we will need to see if that will mess up Dhruv's module, so hold off on enums for now I think. It'll be an improvement we do later

File: Autopilot/SensorFusion/Src/SensorFusion.cpp
Comment (Pop0097) [inline Autopilot/SensorFusion/Src/SensorFusion.cpp:150]: Assuming this is tested and works

File: Autopilot/SensorFusion/Src/SensorFusion.cpp
Comment (Pop0097) [inline Autopilot/SensorFusion/Src/SensorFusion.cpp:428]: I am confused with what is happening here... Is it incomplete?

File: Autopilot/SensorFusion/Src/SensorFusion.cpp
Comment (Pop0097) [inline Autopilot/SensorFusion/Src/SensorFusion.cpp:444]: ditto (though for this one I understand you are still waiting on the driver). 

What I will say tho is this: there really is no need for you to create the `airspeedOutput` struct. Just return the `airspeedData` struct directly.

**Review by Pop0097 — COMMENTED** [review]
My comments so far. Still need to look at the SF_GetPosition method, but I don't think I'll have the time to get to it today...

File: Autopilot/CMakeLists.txt
Comment (Fberbari) [reply Autopilot/CMakeLists.txt]: na, Simulation injects data at the sensorFusion interface. Anything at or lower level than sensor fusion guts are never touched by sim

**Review by Fberbari — APPROVED** [review]
For the moment, my concern is madgwick. And I want this in asap. Everything looks generally OK from my birds eye view though. I'll let Dhruv have the final say, but I'm OK with merging this now, just we can get ends to meet. We can clean up soon after

**Review by Pop0097 — APPROVED** [review]

File: Autopilot/CMakeLists.txt
Comment (Pop0097) [reply Autopilot/CMakeLists.txt]: awesome

**Pull request merged**
